### ISSUE TITLE
CTC loss for Flux

### ISSUE THREAD

**Issue by maetshju** [FluxML/Flux.jl#337]
Of the GSoC work I've done, I think that what's potentially most useful for inclusion in Flux proper is the CTC (connectionist temporal classification) loss function. At the moment, I'm thinking it would fit into the layers section, since that's where the other loss functions are. It seems to me that it would be better off in its own file, since the amount of code for calculating the loss function is quite large compared to your basic mse (for example). Currently, I only have the GPU version working at an acceptable speed.

Ideally, the user-facing interface would look just like the other loss functions, something like `ctc(ŷ, y)`, except that `ŷ` is a sequence, rather than one particular training example.

The GPU code will [look more or less like this](https://github.com/maetshju/gsoc2018/blob/master/speech-cnn/warp-ctc.jl). The interface function is at the bottom of the script.

Alternatively, this could exist in its own package that has easy interop with Flux. Thoughts on whether this would be useful for inclusion in Flux itself?

**Comment by jekbradbury** [comment]
I think it's fine for it to be its own package (CTC.jl, if that's not taken) as long as it works directly with the same user-facing interface as Flux loss functions do (so you'll depend on Flux.jl and use Tracker to wrap the gradient)

**Comment by CarloLucibello** [comment]
I'm no expert of recurrent network, but googling around it seems that CTC is famous enough to deserve inclusion in Flux itself

**Comment by maetshju** [comment]
It has been used by some important speech and handwriting recognition networks for sure, but I do wonder whether it would fit that well with the rest of the Flux code, since there are bare GPU kernels, which are generally rougher to understand than the rest of Flux, in my opinion.

**Comment by MikeInnes** [comment]
Let's just put it in Flux for now, and we can always split it out if it's easier for maintenance or whatever. But otherwise the AD churn is going to be a pain.

We'll need a CPU version if only for testing, it's fine to have something that's not crazy fast. It will also send a clear message of "here's the simple version, here's the fast version".

**Comment by maetshju** [comment]
Opened a PR for CTC, so I'll close this issue.
